joint_state_publisher/joint_state_publisher/joint_state_publisher

```diff
@@ -214,9 +214,14 @@ class JointStatePublisher():
                 elif name in self.dependent_joints:
                     param = self.dependent_joints[name]
                     parent = param['parent']
-                    joint = self.free_joints[parent]
                     factor = param.get('factor', 1)
                     offset = param.get('offset', 0)
+                    while parent in self.dependent_joints:
+                        param = self.dependent_joints[parent]
+                        parent = param['parent']
+                        factor *= param.get('factor', 1)
+                        # the offset is relative only to the first parent
```

Comment:
> Following the formula Ja will move four time as faster as Jc (pretty fine) and it will start in the same location of Jb (2 * 0.8 - 1.6 = 0).

agreed

> Now, I see that it could be a feature for someone, but I would like to be sure that you are aware of this behavior.

I'm aware of the behavior, and it seems sane to me.
I really think it is the only possible interpretation of the URDF spec.

> This seems to me non intuitive starting from the URDF, because it forces me to think of all the interactions within the whole recursive chain.

I have the opposite impression ;)

equation (1) and (2) above are the direct translations of the URDF mimic tags. There is no need to consider the whole chain to understand them and they combine naturally into equation (3).

If we were instead to implement 
```
(3') Ja_value == (x * x') * Jc_value + y'
```
Then, since it is incompatible with equation (2), when reading the urdf we would need to look at Jb definition to know if I need to compute Ja from equation (2) or equation (3').

> A part from the specific example, with the above formulation the starting position of Ja depends not only on its offset, but also from its velocity ratio: 

yes.

> if I set my application and later I need to improve a bit the factor of Ja for a faster motion (x' = 3 in the example), I must also modify its offset to guarantee that the starting position remains unchanged (y' = -2.4 if I want that Ja and Jb start from the same location).

Yes, that seems normal to me too.

The usages of the mimic tags (and the URDF in general) I know of are not to describe behavior but hardware. For instance mimic tags are used for NAO and Pepper hands (a single motor drives all the finger joints) and for NAO HipPitch joints (again two joints driven by a single motor).

I don't know what you're trying to achieve, but it seems out of URDF's (mimic tag) scope.

Would you detail what you are trying to do?

It's probably possible to address your need in joint_state_publisher without changing the interpretation of the URDF files.